We're getting black screens again

### What
- From a user: "The scanner camera does not work. I install them directly... Trash can"
- Device: [Huawei Honor 10](https://play.google.com/console/u/0/developers/4712693179220384697/app/4972942602078310258/user-feedback/reviews#)
Device language: French
App version code: 802
App version name: 3.23.0
Android version: [Android 10 (SDK 29)](https://play.google.com/console/u/0/developers/4712693179220384697/app/4972942602078310258/user-feedback/reviews#)
- From us: "Hello xxx, Are you facing a black screen? Is this a recent install or related to yesterday's update? Does it work again by closing or reopening the app?"
- From the user: "First time installing the app

I installed and then reinstalled black screen.
By closing the app and covering it as you suggest it works.

You should fix this first use BUG. Otherwise there will be a lot of uninstalls

Thank you for responding quickly

Have a good evening"

For the record, this is what I had this morning when fluttering on my old android smartphone:

<details>
<summary>android compileSdkVersion 33 needed</summary>

```
Warning: The plugin url_launcher_android requires Android SDK version 33.

One or more plugins require a higher Android SDK version.
Fix this issue by adding the following to /Users/fabricefontaine/StudioProjects/smooth-app/packages/app/android/app/build.gradle:
android {
  compileSdkVersion 33
  ...
}
```
</details>

<details>
<summary>Finsky crash (whatever that means)</summary>

```
F/Finsky:background(10173): [16299] jzz.h(3): Service object is not present when stopServiceSafely was called.
F/Finsky:background(10173): com.google.android.flib.log.WhatATerribleException
F/Finsky:background(10173): 	at acsv.h(PG:2)
F/Finsky:background(10173): 	at com.google.android.finsky.utils.FinskyLog.k(PG:3)
F/Finsky:background(10173): 	at jzz.h(PG:3)
F/Finsky:background(10173): 	at jzz.g(PG)
F/Finsky:background(10173): 	at jzz.e(PG:2)
F/Finsky:background(10173): 	at kak.h(PG)
F/Finsky:background(10173): 	at kai.a(PG:4)
F/Finsky:background(10173): 	at kak.j(PG:1)
F/Finsky:background(10173): 	at kak.c(PG)
F/Finsky:background(10173): 	at jzw.n(PG:12)
F/Finsky:background(10173): 	at hwx.apply(PG:8)
F/Finsky:background(10173): 	at j$.util.stream.s.accept(SourceFile:45)
F/Finsky:background(10173): 	at j$.util.a.forEachRemaining(SourceFile:20)
F/Finsky:background(10173): 	at j$.util.stream.c.y(SourceFile:21)
F/Finsky:background(10173): 	at j$.util.stream.c.G(SourceFile:8)
F/Finsky:background(10173): 	at j$.util.stream.X0.e(SourceFile)
F/Finsky:background(10173): 	at j$.util.stream.c.J(SourceFile:35)
F/Finsky:background(10173): 	at j$.util.stream.y1.collect(SourceFile:83)
F/Finsky:background(10173): 	at jxm.a(PG:62)
F/Finsky:background(10173): 	at ajpm.e(PG:2)
F/Finsky:background(10173): 	at ajpo.run(PG:9)
F/Finsky:background(10173): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
F/Finsky:background(10173): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
F/Finsky:background(10173): 	at htu.run(PG:15)
F/Finsky:background(10173): 	at java.lang.Thread.run(Thread.java:818)
```
</details>

<details>
<summary>deprecated camera methods</summary>

```
/Users/fabricefontaine/development/flutter/.pub-cache/git/plugins-3c62b56bb6cc0ae1c9594f417ce6305ea45e214f/packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:154: warning: [deprecation] get(int,int) in CamcorderProfile has been deprecated
          return CamcorderProfile.get(cameraId, CamcorderProfile.QUALITY_HIGH);
                                 ^
/Users/fabricefontaine/development/flutter/.pub-cache/git/plugins-3c62b56bb6cc0ae1c9594f417ce6305ea45e214f/packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:158: warning: [deprecation] get(int,int) in CamcorderProfile has been deprecated
          return CamcorderProfile.get(cameraId, CamcorderProfile.QUALITY_2160P);
                                 ^
/Users/fabricefontaine/development/flutter/.pub-cache/git/plugins-3c62b56bb6cc0ae1c9594f417ce6305ea45e214f/packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:162: warning: [deprecation] get(int,int) in CamcorderProfile has been deprecated
          return CamcorderProfile.get(cameraId, CamcorderProfile.QUALITY_1080P);
                                 ^
/Users/fabricefontaine/development/flutter/.pub-cache/git/plugins-3c62b56bb6cc0ae1c9594f417ce6305ea45e214f/packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:166: warning: [deprecation] get(int,int) in CamcorderProfile has been deprecated
          return CamcorderProfile.get(cameraId, CamcorderProfile.QUALITY_720P);
                                 ^
/Users/fabricefontaine/development/flutter/.pub-cache/git/plugins-3c62b56bb6cc0ae1c9594f417ce6305ea45e214f/packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:170: warning: [deprecation] get(int,int) in CamcorderProfile has been deprecated
          return CamcorderProfile.get(cameraId, CamcorderProfile.QUALITY_480P);
                                 ^
/Users/fabricefontaine/development/flutter/.pub-cache/git/plugins-3c62b56bb6cc0ae1c9594f417ce6305ea45e214f/packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:174: warning: [deprecation] get(int,int) in CamcorderProfile has been deprecated
          return CamcorderProfile.get(cameraId, CamcorderProfile.QUALITY_QVGA);
                                 ^
/Users/fabricefontaine/development/flutter/.pub-cache/git/plugins-3c62b56bb6cc0ae1c9594f417ce6305ea45e214f/packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/resolution/ResolutionFeature.java:178: warning: [deprecation] get(int,int) in CamcorderProfile has been deprecated
          return CamcorderProfile.get(cameraId, CamcorderProfile.QUALITY_LOW);
                                 ^
7 warnings
```
</details>

Other:
```
Note: /Users/fabricefontaine/development/flutter/.pub-cache/hosted/pub.dartlang.org/google_mlkit_commons-0.2.0/android/src/main/java/com/google_mlkit_commons/InputImageConverter.java uses unchecked or unsafe operations.
```

At the end, either the camera worked only for a few seconds, or the app crashed, or I had a "black" camera.

--- a/packages/scanner/shared/lib/src/scanner.dart
+++ b/packages/scanner/shared/lib/src/scanner.dart
@@ -43,9 +43,7 @@ abstract class CameraScanner with CameraScannerLogMixin {
   Future<void> onResume() async {}
 
   @mustCallSuper
-  Future<void> onDispose() async {
-    disposeLogs();
-  }
+  Future<void> onDispose() async => disposeLogs();
 
   @protected
   Stream<CameraScannerLog> listenToLogs() => controller;

--- a/packages/smooth_app/lib/pages/scan/ml_kit_scan_page.dart
+++ b/packages/smooth_app/lib/pages/scan/ml_kit_scan_page.dart
@@ -388,9 +388,24 @@ class MLKitScannerPageState extends LifecycleAwareState<MLKitScannerPage>
     await _disposeSoundManager();
   }
 
-  Future<void> _onResumeImageStream({bool forceStartPreview = false}) async {
+  /// [shouldRetry] indicates if we should retry to resume if the screen is invisible
+  Future<void> _onResumeImageStream({
+    bool forceStartPreview = false,
+    bool shouldRetry = true,
+  }) async {
     if (stoppingCamera ||
         (!forceStartPreview && ScreenVisibilityDetector.invisible(context))) {
+      if (shouldRetry) {
+        SchedulerBinding.instance.addPostFrameCallback((_) {
+          _onResumeImageStream(
+            forceStartPreview: forceStartPreview,
+            shouldRetry: false,
+          );
+        });
+
+        _redrawScreen();
+      }
+
       return;
     }
 
@@ -412,7 +427,7 @@ class MLKitScannerPageState extends LifecycleAwareState<MLKitScannerPage>
         // Just wait
         return;
       }
-      return _stopImageStream();
+      return _stopImageStream(autoRestart: true);
     }
 
     if (_streamSubscription?.isPaused == true) {
@@ -430,6 +445,7 @@ class MLKitScannerPageState extends LifecycleAwareState<MLKitScannerPage>
         return _stopImageStream();
       }
     }
+
     stoppingCamera = false;
   }
 
@@ -454,7 +470,9 @@ class MLKitScannerPageState extends LifecycleAwareState<MLKitScannerPage>
     CameraHelper.destroyControllerInstance();
 
     await _streamSubscription?.cancel();
-    await _barcodeDecoder.onDispose();
+    try {
+      await _barcodeDecoder.onDispose().timeout(const Duration(seconds: 5));
+    } catch (_) {}
 
     stoppingCamera = false;
 

--- a/packages/smooth_app/lib/widgets/screen_visibility.dart
+++ b/packages/smooth_app/lib/widgets/screen_visibility.dart
@@ -25,10 +25,16 @@ class _ScreenVisibilityDetectorState extends State<ScreenVisibilityDetector> {
 
   @override
   Widget build(BuildContext context) {
+    // The first time the Widget is called, [onVisibilityChanged] is not called
+    // And it will keep the initial value -> [false] which is wrong
+    if (!_notifier.isVisible) {
+      _notifier.updateValue(true);
+    }
+
     return VisibilityDetector(
       key: const ValueKey<String>('ScreenVisibility'),
       onVisibilityChanged: (VisibilityInfo info) {
-        _notifier.updateValue(info);
+        _notifier.updateValue(info.visible);
       },
       child: ChangeNotifierProvider<ScreenVisibility>.value(
         value: _notifier,
@@ -43,9 +49,7 @@ class ScreenVisibility extends ValueNotifier<bool> {
 
   bool get isVisible => value;
 
-  void updateValue(VisibilityInfo info) {
-    final bool visible = info.visible;
-
+  void updateValue(bool visible) {
     if (value != visible) {
       value = visible;
     }

